src/build/app/mod.rs

```diff
@@ -236,6 +236,73 @@ impl<'help> App<'help> {
         self.get_opts().filter(|a| a.get_help_heading().is_none())
     }
 
+    /// Test if the provided Argument is a global Argument
+    pub fn test_global(arg: &Arg) -> bool {
+        arg.global
+    }
+
+    /// Get a list of subcommands which contain the provided Argument
+    ///
+    /// This command will only include subcommands in its list for which the subcommands
+    /// parent also contains the Argument.
+    ///
+    /// **NOTE:** In this case only Sucommand_1 will be included
+    ///   Subcommand_1 (contains Arg)
+    ///     Subcommand_1.1 (doesn't contain Arg)
+    ///       Subcommand_1.1.1 (contains Arg)
+    ///
+    pub fn get_subcommands_containing(&self, arg: &Arg) -> Vec<&App<'help>> {
+        let mut vec = std::vec::Vec::new();
+        for idx in 0..self.subcommands.len() {
+            if self.subcommands[idx]
+                .args
+                .args
+                .iter()
+                .any(|ar| ar.id == arg.id)
+            {
+                vec.push(&self.subcommands[idx]);
+                // TODO rustc >= 1.46: change  .get(dix).unwrap()  to [idx]
+                // improved readability
+                vec.append(
+                    &mut self
+                        .subcommands
+                        .get(idx)
+                        .unwrap()
+                        .get_subcommands_containing(arg),
+                );
+            }
+        }
+        vec
+    }
+
+    /// Get a unique list of all arguments of all commands and continuous subcommands the given argument conflicts with.
+    ///
+    /// ### Panics
+    ///
+    /// If the given arg contains a conflict with an argument that is unknown to
+    /// this `App`.
+    pub fn get_global_arg_conflicts_with(&self, arg: &Arg) -> Vec<&Arg<'help>> // FIXME: This could probably have been an iterator
+    {
+        arg.blacklist
+            .iter()
+            .map(|id| {
+                self.args
+                    .args
+                    .iter()
+                    .chain(
+                        self.get_subcommands_containing(arg)
+                            .iter()
+                            .flat_map(|x| x.args.args.iter()),
+                    )
+                    .find(|arg| arg.id == *id)
+                    .expect(
```

Comment:
I am not sure why we are panicking here.

Reply:
The panic here is the same behavior that the non-global version of the method has.
This would be consistent with the implementation of get_arg_conflicts_with that was already there.

Reply:
Let's move both the functions you added here into zsh.

Reply:
This method is part of the App. When I move them I don't have access to the internals.

Reply:
The major issue here is `get_subcommands_containing`. It needs better doc (and it should be private) for the code to be readable and maintainable.

Reply:
When it's private I can't access it from clap_generate. I could write a longer explanation in the docs but that would include a description of how the global status of an argument effects its parents and children.

Reply:
Keep this one public but the other private (with added docs because it is very specific to this situation)

Reply:
Ok